server/src/test/java/org/opensearch/tasks/TaskResourceTrackingServiceTests.java

```diff
@@ -88,6 +93,48 @@ public void testStopTrackingHandlesCurrentActiveThread() {
         assertTrue(task.getResourceStats().get(threadId).get(0).getResourceUsageInfo().getStatsInfo().get(MEMORY).getTotalValue() > 0);
     }
 
+    /**
+     * Test if taskResourceTrackingService properly tracks resource usage when multiple threads work on the same task
+     */
+    public void testStartingTrackingHandlesMultipleThreadsPerTask() throws InterruptedException {
+        ExecutorService executor = threadPool.executor(ThreadPool.Names.GENERIC);
+        taskResourceTrackingService.setTaskResourceTrackingEnabled(true);
+        Task task = new SearchTask(1, "test", "test", () -> "Test", TaskId.EMPTY_TASK_ID, new HashMap<>());
+        taskResourceTrackingService.startTracking(task);
+        int numTasks = randomIntBetween(2, 100);
+        for (int i = 0; i < numTasks; i++) {
+            executor.execute(() -> {
+                long threadId = Thread.currentThread().getId();
+                taskResourceTrackingService.taskExecutionStartedOnThread(task.getId(), threadId);
+                // The same thread may pick up multiple runnables for the same task id
+                assertEquals(1, task.getResourceStats().get(threadId).stream().filter(ThreadResourceInfo::isActive).count());
+                taskResourceTrackingService.taskExecutionFinishedOnThread(task.getId(), threadId);
+            });
+        }
+        executor.shutdown();
+        while (true) {
+            try {
+                if (executor.awaitTermination(1, TimeUnit.MINUTES)) break;
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        Map<Long, List<ThreadResourceInfo>> stats = task.getResourceStats();
+        int numExecutions = 0;
+        for (Long threadId : stats.keySet()) {
+            for (ThreadResourceInfo info : task.getResourceStats().get(threadId)) {
+                assertTrue(info.getResourceUsageInfo().getStatsInfo().get(MEMORY).getTotalValue() > 0);
+                assertTrue(info.getResourceUsageInfo().getStatsInfo().get(CPU).getTotalValue() > 0);
```

Comment:
should we add assert that non of the thread is active now ?

Reply:
Good call, added